File: app/src/main/java/com/sankar/kotlinmvvm/view/ListViewActivity.java
package com.sankar.kotlinmvvm.view;

import android.os.Bundle;

import android.view.View;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.sankar.kotlinmvvm.R;
import com.sankar.kotlinmvvm.adapter.ListViewAdapter;
import com.sankar.kotlinmvvm.model.ListModel;
import com.sankar.kotlinmvvm.response.ListViewResponse;
import com.sankar.kotlinmvvm.viewmodel.ListViewViewModel;

import java.util.ArrayList;
import java.util.List;

public class ListViewActivity extends AppCompatActivity {

    private static final String TAG = ListViewActivity.class.getSimpleName();
    private RecyclerView my_recycler_view;
    private ProgressBar progress_circular_movie_article;
    private LinearLayoutManager layoutManager;
    private ListViewAdapter adapter;
    private ArrayList<ListModel> articleArrayList = new ArrayList<>();
    ListViewViewModel articleViewModel;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initialization();

        getMovieArticles();
    }

    /**
     * initialization of views and others
     *
     * @param @null
     */
    private void initialization() {
        progress_circular_movie_article = (ProgressBar) findViewById(R.id.progress_circular_movie_article);
        my_recycler_view = (RecyclerView) findViewById(R.id.my_recycler_view);

        // use a linear layout manager
        layoutManager = new LinearLayoutManager(ListViewActivity.this);
        my_recycler_view.setLayoutManager(layoutManager);

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        my_recycler_view.setHasFixedSize(true);

        // adapter
        adapter = new ListViewAdapter(ListViewActivity.this, articleArrayList);
        my_recycler_view.setAdapter(adapter);

        // View Model
        articleViewModel = ViewModelProviders.of(this).get(ListViewViewModel.class);
    }

    /**
     * get movies articles from news api
     *
     * @param @null
     */
    private void getMovieArticles() {
        articleViewModel.getArticleResponseLiveData().observe(this, new Observer<ListViewResponse>() {
            @Override
            public void onChanged(ListViewResponse articleResponse) {
                if (articleResponse != null) {

                    progress_circular_movie_article.setVisibility(View.GONE);
                    List<ListModel> articles = articleResponse.getArticles();
                    articleArrayList.addAll(articles);
                    adapter.notifyDataSetChanged();
                }
            }
        });
    }
}

File: app/src/main/java/com/sankar/kotlinmvvm/view/ListViewActivity.kt
package com.sankar.kotlinmvvm.view

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.sankar.kotlinmvvm.R
import com.sankar.kotlinmvvm.adapter.ListViewAdapter
import com.sankar.kotlinmvvm.model.Model
import com.sankar.kotlinmvvm.viewmodel.ListViewViewModel
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class ListViewActivity : AppCompatActivity() {
    private var adapter: ListViewAdapter? = null
    private val articleArrayList =
        ArrayList<Model.Article>()
    var articleViewModel: ListViewViewModel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initialization()
        movieArticles
    }

    private fun initialization() {
        my_recycler_view!!.layoutManager = LinearLayoutManager(this@ListViewActivity)
        adapter = ListViewAdapter(this@ListViewActivity, articleArrayList)
        my_recycler_view!!.adapter = adapter

        articleViewModel =
            ViewModelProviders.of(this).get(ListViewViewModel::class.java)
    }

    private val movieArticles: Unit
        private get() {
            articleViewModel!!.articleResponseLiveData
                .observe(this,
                    Observer { articleResponse ->
                        if (articleResponse != null) {
                            progress_circular_movie_article!!.visibility = View.GONE
                            val articles: List<Model.Article>? =
                                articleResponse.articles
                            if (articles != null) {
                                articleArrayList.addAll(articles)
                            }
                            adapter!!.notifyDataSetChanged()
                        }
                    })
        }

    companion object {
        private val TAG = ListViewActivity::class.java.simpleName
    }
}
